File: typescript/packages/core/tests/failure-step-span.test.ts
import { expect, test } from 'vitest'
import { executePlan } from '../src/execute.ts'
import { toFailure } from '../src/failure.ts'
import { attachFailureAnchor, readFailureAnchor } from '../src/failure-anchor.ts'
import { hashSource } from '../src/hash.ts'
import { parse } from '../src/parse.ts'
import { plan } from '../src/plan.ts'
import { addStep, createRegistry } from '../src/registry.ts'
import { runResultDiagnostics } from '../src/run-diagnostics.ts'
import { spanFromOffsets } from '../src/span.ts'

// A throwing sensor sharing its line with a stimulus that passed: the whole
// chain (executePlan → toFailure → runResultDiagnostics) must land on the
// sensor's own text, since underlining the line would blame the stimulus too.
const SOURCE = '# L\n\nHe asks on June 10, and the library agrees.\n'
const STEP_TEXT = 'the library agrees'

function throwingPlan() {
  let r = createRegistry()
  r = addStep(r, {
    expression: 'asks on June 10',
    expressionSourceFile: 's.ts',
    expressionSourceLine: 1,
    kind: 'stimulus',
    handler: () => {},
  })
  r = addStep(r, {
    expression: STEP_TEXT,
    expressionSourceFile: 's.ts',
    expressionSourceLine: 2,
    kind: 'sensor',
    handler: () => {
      throw new Error('expected the library to refuse')
    },
  })
  return plan(parse('l.md', SOURCE), r)
}

async function failureOf() {
  const p = throwingPlan()
  let run: (() => void | Promise<void>) | undefined
  executePlan(p, {
    sink: {
      example: (_n, r) => {
        run = r
      },
    },
    reporter: { diagnostic: () => {} },
  })
  try {
    await run?.()
  } catch (error) {
    return toFailure(error, 'l.md', 3)
  }
  throw new Error('the example was expected to fail')
}

test('a thrown step records the anchor of the step that threw', async () => {
  const f = await failureOf()
  expect(f.anchor).toBeDefined()
  expect(SOURCE.slice(f.anchor?.from ?? 0, f.anchor?.to ?? 0)).toBe(STEP_TEXT)
})

test('the diagnostic underlines the failing step, leaving the passing one alone', async () => {
  const f = await failureOf()
  const diags = runResultDiagnostics(
    {
      version: 1,
      oathPath: 'l.md',
      sourceHash: hashSource(SOURCE),
      examples: [{ name: 'e', status: 'failed', lines: [3], failure: f }],
    },
    SOURCE,
  )
  expect(diags).toHaveLength(1)
  expect(SOURCE.slice(diags[0]?.from ?? 0, diags[0]?.to ?? 0)).toBe(STEP_TEXT)
  expect(diags[0]?.message).toBe('expected the library to refuse')
})

test('the anchor rides on the error without becoming an enumerable property', () => {
  const err = new Error('boom')
  attachFailureAnchor(err, spanFromOffsets(SOURCE, 5, 9))
  expect(readFailureAnchor(err)?.startOffset).toBe(5)
  expect(JSON.stringify({ ...err })).toBe('{}')
  // Non-objects are ignored rather than throwing — a step may throw anything.
  expect(() => attachFailureAnchor('a string', spanFromOffsets(SOURCE, 0, 1))).not.toThrow()
  expect(readFailureAnchor('a string')).toBeUndefined()
})

File: rust/core/tests/failure_step_span_test.rs
//! Port of `failure-step-span.test.ts`. A failing sensor sharing its line with a
//! stimulus that passed: the chain (executor → [`to_failure`]) must land on the
//! sensor's own text, since a renderer underlining the line would blame the
//! stimulus too.

use varar_core::error::StepError;
use varar_core::execute::{ExecutePorts, collect_examples};
use varar_core::failure::to_failure;
use varar_core::handler::Handler;
use varar_core::offsets::utf16_slice;
use varar_core::parse::parse;
use varar_core::plan::plan;
use varar_core::registry::{add_step, create_registry};
use varar_core::step_kind::StepKind;

const SOURCE: &str = "# L\n\nHe asks on June 10, and the library agrees.\n";
const STEP_TEXT: &str = "the library agrees";

#[test]
fn a_failing_step_records_the_anchor_of_the_step_that_failed() {
    let mut r = create_registry();
    r = add_step(
        &r,
        "asks on June 10",
        "s.rs",
        1,
        Handler::sync0(|_s| Ok(None)),
        Some(StepKind::Stimulus),
    )
    .unwrap();
    r = add_step(
        &r,
        STEP_TEXT,
        "s.rs",
        2,
        Handler::sync0(|_s| panic!("expected the library to refuse")),
        Some(StepKind::Sensor),
    )
    .unwrap();

    let p = plan(&parse("l.md", SOURCE), &r);
    let ports = ExecutePorts::silent();
    let failure = collect_examples(&p, &ports)[0].run().unwrap_err();
    assert!(matches!(failure.error, StepError::Handler(_)));

    let f = to_failure(&failure, "l.md", 3);
    let anchor = f.anchor.expect("a failing step records an anchor");
    assert_eq!(STEP_TEXT, utf16_slice(SOURCE, anchor.from, anchor.to));
}

#[test]
fn a_failure_with_no_location_for_this_oath_has_no_anchor() {
    let sf = varar_core::error::StepFailure::bare(StepError::ReturnShape("nope".to_string()));
    assert_eq!(None, to_failure(&sf, "l.md", 7).anchor);
}
